clang/lib/CIR/Dialect/IR/CIRAttrs.cpp

```diff
@@ -295,6 +295,89 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// DataMemberPtrAttr definitions
```

Comment:
What I'm thinking is that there's just too much differences between a ptr-to-data-member and a ptr-to-member-function, so it seems not reasonable enough to invent a general `#cir.member_ptr` attribute to represent both of them.

We can use a single `cir.member_ptr` type to represent both ptr-to-data-member type and ptr-to-function-member type because we can distinguish these two by simply testing whether the member type is a function type. However, while a ptr-to-data-member constant can be represented as a `(cls_ty, member_index)` pair, a ptr-to-function-member constant cannot be represented in a similar form. At least you need to pack a function pointer in it, and in the case of virtual member functions things are even more complex.

Reply:
> At least you need to pack a function pointer in it, and in the case of virtual member functions things are even more complex.

Good point!